File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':

        new_head = Node(0)
        if head is None:
            return head

        temp = new_head
        curr = head
        d = {}
        
        while curr:
            temp.val = curr.val
            d[curr] = temp
            if curr.next:
                temp.next = Node(0)
            temp = temp.next
            curr = curr.next
        
        curr = head
        temp = new_head

        while curr:
            if curr.random:
                temp.random = d[curr.random]
            curr= curr.next
            temp = temp.next

            
        return new_head
            

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
         if(lists.length==0) return null;

         PriorityQueue<ListNode> minHeap = new PriorityQueue<>((a,b) -> a.val-b.val);


         for(ListNode head : lists)
         {
            minHeap.offer(head);
         }

         ListNode res=new ListNode(0);
         ListNode head=res;

         while(!minHeap.isEmpty())
         {
            ListNode node=minHeap.poll();
            head.next=node;
            head=head.next;

            node=node.next;
            if(node!=null) minHeap.offer(node);
        
         }

         return res.next;
    }
}
